lib/Sema/TypeCheckPattern.cpp

```diff
@@ -1556,6 +1557,28 @@ bool TypeChecker::coerceParameterListToType(ParameterList *P, DeclContext *DC,
     return hadError;
   };
 
+  if (!hadError && isa<ParenType>(paramListType.getPointer())) {
```

Comment:
I got what this was doing after a few read-throughs.  Could you document the flow here?